clash-lib/src/Clash/Core/Evaluator/Models.hs

```diff
@@ -0,0 +1,258 @@
+deleteEnv scope i env =
+  case scope of
+    LocalId -> env { envLocals = Map.delete i (envLocals env) }
+    GlobalId -> env { envGlobals = delVarEnv (envGlobals env) i }
+
+updateEnvPrim :: Int -> Value -> Env -> Env
+updateEnvPrim i v env =
+  env { envPrimsIO = (IntMap.insert i v pm, n) }
+ where
+  (pm, n) = envPrimsIO env
+
+-- | Neutral terms cannot be reduced, as they represent things like variables
+-- which are unknown, partially applied functions, or case expressions where
+-- the scrutinee is not yet an inspectable value. Consider:
+--
+-- v              Stuck if "v" is a free variable
+-- p x1 .. xn     Stuck if "p" has no primitive reduction rule
+-- x $ y          Stuck if "x" is not known to be a lambda
+-- x @ A          Stuck if "x" is not known to be a type lambda
+-- case x of ...  Stuck if "x" is neutral (cannot choose an alternative)
+--
+data Neutral a
+  = NeVar   (Var Term)
+  | NePrim  PrimInfo [Either Value Type]
+  | NeApp   (Neutral a) a
+  | NeTyApp (Neutral a) Type
+  | NeCase  a Type [(Pat, a)]
+  deriving (Show, Generic, NFData)
+
+-- A term which has been normalised to weak head normal form (WHNF). This has
+-- no redexes at the head of the term, but subterms may still contain redexes.
+--
+data Value
+  = VNeu    (Neutral Value)
+  | VData   DataCon [Either Value Type]
+  | VLit    Literal
+  | VPrim   PrimInfo [Either Value Type]
+  | VLam    Id Term Env
+  | VTyLam  TyVar Term Env
+  | VCast   Value Type Type
+  | VTick   Value TickInfo
+  deriving (Show, Generic, NFData)
+
+collectValueTicks :: Value -> (Value, [TickInfo])
+collectValueTicks = go []
+ where
+  go acc (VTick v ti) = go (ti:acc) v
+  go acc v = (v, acc)
+
+addTicks :: Value -> [TickInfo] -> Value
+addTicks = foldl' VTick
+
+-- | A term which is in beta-long eta-normal (NF). This has no redexes, and all
+-- partially applied functions in subterms are eta-expanded.
+--
+data Nf
+  = NNeu    (Neutral Nf)
+  | NData   DataCon [Either Nf Type]
+  | NLit    Literal
+  | NPrim   PrimInfo [Either Nf Type]
+  | NLam    Id Nf
+  | NTyLam  TyVar Nf
+  | NCast   Nf Type Type
+  | NTick   Nf TickInfo
+  deriving (Show, Generic, NFData)
+
+-- Embedding WHNF and HNF values back into Term.
+--
+class AsTerm a where
+  asTerm :: a -> Term
+
+instance AsTerm Term where
+  asTerm = id
+
+instance (AsTerm a) => AsTerm (Neutral a) where
+  asTerm (NeVar v)        = Var v
+  asTerm (NePrim p args)  = mkApps (Prim p) (first asTerm <$> args)
+  asTerm (NeApp x y)      = App (asTerm x) (asTerm y)
+  asTerm (NeTyApp x ty)   = TyApp (asTerm x) ty
+  asTerm (NeCase x ty as) = Case (asTerm x) ty (second asTerm <$> as)
+
+instance AsTerm Value where
+  asTerm (VNeu n)         = asTerm n
+  asTerm (VData dc args)  = mkApps (Data dc) (first asTerm <$> args)
+  asTerm (VLit l)         = Literal l
+  asTerm (VPrim p args)   = mkApps (Prim p) (first asTerm <$> args)
+  asTerm (VLam x e env)   = substTm "asTerm.VLam" (asSubst env) (Lam x e)
```

Comment:
Why not substitute the types of the environment, and let-bind the terms? instead substituting both.

Reply:
That's actually something I thought about yesterday, and one of the things me and @martijnbastiaan talked about on our call this morning (reaching the same conclusion) :+1:

One extra thing I've noticed though: The old evaluator did eagerly substitute terms when unwinding an `Apply` strack frame. Should these substitutions still occur, and if so, how do you decide between terms that should be in a `let` binding, or terms that should be substituted?

Reply:
That's because we only have variables on the `Apply` stack frame, so no sharing is lost.